src/ucp/wireup/select.c

```diff
@@ -243,31 +238,44 @@ static int ucp_wireup_check_amo_flags(const uct_tl_resource_desc_t *resource,
 
 static int
 ucp_wireup_check_keepalive(const ucp_wireup_select_params_t *select_params,
-                           const uct_tl_resource_desc_t *resource,
-                           uint64_t flags, uint64_t required_flags,
-                           const char *title, const char **flag_descs,
+                           ucp_rsc_index_t rsc_index, uint64_t flags,
+                           uint64_t required_flags, const char *title,
+                           int is_keepalive, const char **flag_descs,
                            char *reason, size_t max)
 {
-    ucp_worker_h worker = select_params->ep->worker;
+    ucp_worker_h worker                    = select_params->ep->worker;
+    ucp_context_h context                  = worker->context;
+    const uct_tl_resource_desc_t *resource = &context->tl_rscs[rsc_index].tl_rsc;
     char title_keepalive[128];
     char title_ep_check[128];
+    char title_am_based[128];
+
+    if (!is_keepalive) {
+        /* Keepalive is not needed */
+        return 1;
+    }
 
     ucs_snprintf_safe(title_keepalive, sizeof(title_keepalive),
                       "%s with keepalive", title);
     ucs_snprintf_safe(title_ep_check, sizeof(title_ep_check),
                       "%s with ep_check", title);
-    return /* if error handling and keepalive were requested, UCT iface has to
-            * support peer failure (i.e. UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE)
-            * and either built-in keepalive (i.e. UCT_IFACE_FLAG_EP_KEEPALIVE)
-            * or EP checking (i.e. UCT_IFACE_FLAG_EP_CHECK) */
-            !ucp_worker_keepalive_is_enabled(worker) ||
-            !(select_params->ep_init_flags &
-              UCP_EP_INIT_ERR_MODE_PEER_FAILURE) ||
-            ucp_wireup_check_flags(resource, flags, UCT_IFACE_FLAG_EP_KEEPALIVE,
+    ucs_snprintf_safe(title_am_based, sizeof(title_am_based),
+                      "%s with am-based keepalive", title);
+
+    return /* Either built-in keepalive (i.e. UCT_IFACE_FLAG_EP_KEEPALIVE) or
+            * EP checking (i.e. UCT_IFACE_FLAG_EP_CHECK) or CONNECT_TO_IFACE
+            * and AM_BCOPY for AM-based keepalive */
+            ucp_wireup_check_flags(resource, flags,
+                                   UCT_IFACE_FLAG_EP_KEEPALIVE,
                                    title_keepalive, ucp_wireup_iface_flags,
                                    reason, max) ||
             ucp_wireup_check_flags(resource, flags, UCT_IFACE_FLAG_EP_CHECK,
```

Comment:
shall we also add `CONNECT_TO_EP` for this check?

Reply:
good catch, done